File: backend/app/schemas/lesson_share.py
"""
教案分享 API Schema - AI英语教学系统

定义教案分享相关的请求和响应数据结构。
"""
import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class SharePermission(str, Enum):
    """分享权限枚举"""
    VIEW = "view"      # 仅查看
    EDIT = "edit"      # 可编辑
    COPY = "copy"      # 可复制


class ShareStatus(str, Enum):
    """分享状态枚举"""
    PENDING = "pending"    # 待接受
    ACCEPTED = "accepted"  # 已接受
    REJECTED = "rejected"  # 已拒绝
    EXPIRED = "expired"    # 已过期


# ==================== 请求 Schema ====================

class CreateShareRequest(BaseModel):
    """创建分享请求"""
    shared_to: uuid.UUID = Field(..., description="接收分享的教师ID")
    permission: SharePermission = Field(
        default=SharePermission.VIEW,
        description="分享权限"
    )
    message: str | None = Field(None, max_length=500, description="分享附言")
    expires_days: int | None = Field(
        None,
        ge=1,
        le=365,
        description="有效期（天数），不设置则永久有效"
    )


class UpdateShareStatusRequest(BaseModel):
    """更新分享状态请求"""
    status: ShareStatus = Field(..., description="新状态")


# ==================== 响应 Schema ====================

class UserSummary(BaseModel):
    """用户摘要信息"""
    id: uuid.UUID
    username: str
    full_name: str | None
    email: str

    class Config:
        from_attributes = True


class LessonPlanSummary(BaseModel):
    """教案摘要信息"""
    id: uuid.UUID
    title: str
    topic: str
    level: str
    duration: int
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ShareResponse(BaseModel):
    """分享记录响应"""
    id: uuid.UUID
    lesson_plan: LessonPlanSummary
    shared_by: UserSummary
    shared_to: UserSummary
    permission: SharePermission
    status: ShareStatus
    message: str | None
    expires_at: datetime | None
    created_at: datetime

    class Config:
        from_attributes = True


class ShareListResponse(BaseModel):
    """分享列表响应"""
    shares: list[ShareResponse]
    total: int
    page: int
    page_size: int


class CreateShareResponse(BaseModel):
    """创建分享响应"""
    share: ShareResponse
    message: str = "分享已发送，等待对方接受"


# ==================== 复制相关 Schema ====================

class DuplicateLessonPlanRequest(BaseModel):
    """复制教案请求"""
    new_title: str | None = Field(None, max_length=255, description="新教案标题，不设置则自动添加副本标识")
    include_shares: bool = Field(
        default=False,
        description="是否同时复制分享记录（通常不复制）"
    )


class DuplicateLessonPlanResponse(BaseModel):
    """复制教案响应"""
    lesson_plan: LessonPlanSummary
    message: str = "教案已成功复制"


class CreateFromTemplateRequest(BaseModel):
    """从模板创建教案请求"""
    title: str = Field(..., max_length=255, description="教案标题")
    topic: str = Field(..., max_length=255, description="教学主题")
    level: str = Field(..., pattern=r'^(A1|A2|B1|B2|C1|C2)$', description="CEFR等级")
    duration: int = Field(default=45, ge=15, le=180, description="课程时长（分钟）")
    target_exam: str | None = Field(None, max_length=50, description="目标考试类型")
    additional_requirements: str | None = Field(
        None,
        max_length=1000,
        description="额外要求"
    )


class LessonPlanTemplateSummary(BaseModel):
    """教案模板摘要"""
    id: uuid.UUID
    name: str
    description: str | None
    level: str
    target_exam: str | None
    is_system: bool
    usage_count: int
    created_at: datetime

    class Config:
        from_attributes = True


class TemplateListResponse(BaseModel):
    """模板列表响应"""
    templates: list[LessonPlanTemplateSummary]
    total: int
    page: int
    page_size: int

File: frontend/src/api/lessonShare.ts
/**
 * 教案分享 API 客户端
 *
 * 提供教案分享相关的API调用方法。
 */
import request from '@/utils/request'

export interface SharePermission {
  value: 'view' | 'edit' | 'copy'
  label: string
}

export interface ShareStatus {
  value: 'pending' | 'accepted' | 'rejected' | 'expired'
  label: string
}

// 分享权限枚举
export const SHARE_PERMISSIONS: SharePermission[] = [
  { value: 'view', label: '仅查看' },
  { value: 'edit', label: '可编辑' },
  { value: 'copy', label: '可复制' }
]

// 分享状态枚举
export const SHARE_STATUSES: ShareStatus[] = [
  { value: 'pending', label: '待接受' },
  { value: 'accepted', label: '已接受' },
  { value: 'rejected', label: '已拒绝' },
  { value: 'expired', label: '已过期' }
]

export interface UserSummary {
  id: string
  username: string
  full_name: string | null
  email: string
}

export interface LessonPlanSummary {
  id: string
  title: string
  topic: string
  level: string
  duration: number
  status: string
  created_at: string
  updated_at: string
}

export interface LessonPlanShare {
  id: string
  lesson_plan: LessonPlanSummary
  shared_by: UserSummary
  shared_to: UserSummary
  permission: string
  status: string
  message: string | null
  expires_at: string | null
  created_at: string
}

export interface CreateShareRequest {
  shared_to: string
  permission: string
  message?: string
  expires_days?: number
}

export interface ShareListResponse {
  shares: LessonPlanShare[]
  total: number
  page: number
  page_size: number
}

export interface DuplicateLessonPlanRequest {
  new_title?: string
  include_shares?: boolean
}

export interface DuplicateLessonPlanResponse {
  lesson_plan: LessonPlanSummary
  message: string
}

export interface CreateFromTemplateRequest {
  title: string
  topic: string
  level: string
  duration?: number
  target_exam?: string
  additional_requirements?: string
}

export interface LessonPlanTemplateSummary {
  id: string
  name: string
  description: string | null
  level: string
  target_exam: string | null
  is_system: boolean
  usage_count: number
  created_at: string
}

export interface TemplateListResponse {
  templates: LessonPlanTemplateSummary[]
  total: number
  page: number
  page_size: number
}

/**
 * 分享教案给其他教师
 */
export function shareLessonPlan(lessonPlanId: string, data: CreateShareRequest) {
  return request<LessonPlanShare>({
    url: `/api/v1/lesson-plans/${lessonPlanId}/share`,
    method: 'post',
    data
  })
}

/**
 * 获取分享给我的教案列表
 */
export function getSharedWithMe(params?: {
  status?: string
  page?: number
  page_size?: number
}) {
  return request<ShareListResponse>({
    url: '/api/v1/lesson-plans/shared',
    method: 'get',
    params
  })
}

/**
 * 获取我分享的教案列表
 */
export function getSharedByMe(params?: {
  status?: string
  page?: number
  page_size?: number
}) {
  return request<ShareListResponse>({
    url: '/api/v1/lesson-plans/shared-by-me',
    method: 'get',
    params
  })
}

/**
 * 接受分享
 */
export function acceptShare(shareId: string) {
  return request<LessonPlanShare>({
    url: `/api/v1/lesson-plans/shared/${shareId}/accept`,
    method: 'put'
  })
}

/**
 * 拒绝分享
 */
export function rejectShare(shareId: string) {
  return request<LessonPlanShare>({
    url: `/api/v1/lesson-plans/shared/${shareId}/reject`,
    method: 'put'
  })
}

/**
 * 取消分享
 */
export function cancelShare(shareId: string) {
  return request({
    url: `/api/v1/lesson-plans/shared/${shareId}`,
    method: 'delete'
  })
}

/**
 * 复制教案
 */
export function duplicateLessonPlan(lessonPlanId: string, data?: DuplicateLessonPlanRequest) {
  return request<DuplicateLessonPlanResponse>({
    url: `/api/v1/lesson-plans/${lessonPlanId}/duplicate`,
    method: 'post',
    data
  })
}

/**
 * 获取教案模板列表
 */
export function getLessonPlanTemplates(params?: {
  level?: string
  target_exam?: string
  is_system?: boolean
  page?: number
  page_size?: number
}) {
  return request<TemplateListResponse>({
    url: '/api/v1/lesson-plans/templates',
    method: 'get',
    params
  })
}

/**
 * 从模板创建教案
 */
export function createFromTemplate(templateId: string, data: CreateFromTemplateRequest) {
  return request<{
    lesson_plan: LessonPlanSummary
  }>({
    url: `/api/v1/lesson-plans/from-template/${templateId}`,
    method: 'post',
    data
  })
}
